### DISCUSSION TITLE
Xstate test: How to control service return or error depending upon test plan.

### DISCUSSION THREAD

**Discussion by DavidOrchard** [statelyai/xstate#2911] [Q&A]
**Description**

We'd like to be able to model based test xstate services.  The next state, ie success or failure, is dependent upon a service return.  We like to have a callback in the test run that can look at which path it's on, ie success or failure test path, and cause an effect on the service such that it returns the value needed.  In our real world application, we do API calls in services which are mocked with msw and we configure the mock response based on the test path.

**Expected Result**

Both success and failure test paths are successfully executed

**Actual Result**

Only one of the test paths is executed and it can't be configured at test run time.

**Reproduction**

- https://codesandbox.io/s/affectionate-goldberg-92ydo

**Additional context**

**Comment by davidkpiano** [comment]
For now, you can specify the `done.invoke` and `error.invoke` events for the invocation:

```js
const testModel = createModel(...).withEvents({
  'done.invoke.loading': () => {
    // mock success
  },
  'error.invoke.loading': () => {
    // mock error
  }
});

```

Try that out

**Reply by DavidOrchard** [reply]
I figured out a different way!  I can check the target state in the test and affect the loading service.  I've updated the sandbox to show this.
```
          value = Object.keys(path.state.meta)[0] === "statemachine.success";
```

I'll try the event approach too.

Thank you for the very quick & helpful response. It gives me even more faith about the ongoing maintenance of these repos.

**Reply by ndthanhdev** [reply]
> For now, you can specify the `done.invoke` and `error.invoke` events for the invocation:
> 
> ```js
> const testModel = createModel(...).withEvents({
>   'done.invoke.loading': () => {
>     // mock success
>   },
>   'error.invoke.loading': () => {
>     // mock error
>   }
> });
> ```
> 
> Try that out

I tried the above approach, but it didn't work because the get request had started before the mock was applied.

Could you add a way to apply mock before events? Ex: a hook `before.error.invoke.fetching`, or always add a small delay before each invoke...

```ts
[`error.invoke.fetching`]: {
  exec: async (ctx) => {
	  ctx.mswServer.use(
		  rest.get(
			  "https://my-json-server.typicode.com/typicode/demo/posts/1",
			  (req, res, ctx) => {
				  return res(ctx.status(400));
			  }
		  )
	  );
  },
},
```
